README.md
# MaUde

**M**etric for **a**utomatic **U**nreferenced **d**ialog **e**valuation.

Contains code of the paper titled _"Learning an Unreferenced Metric for Online Dialogue Evaluation"_ to appear at **ACL 2020**, [Arxiv](https://arxiv.org/abs/2005.00583)

## Getting the data

To get the trained models, [download the data from here](https://drive.google.com/file/d/1Ysso9hdzSenK13LjOFombyXYqA_kv-Vy/view?usp=sharing).

## Computing Backtranslation

We use [FairSeq](https://github.com/pytorch/fairseq) to compute back-translations. Our modified scripts are present in `scripts` folder, to run cd into that folder and run `./run_trans.sh`.  

## Computing Corruption Files

In the data dump we already provide the corruption files used for training. To generate new corruption files on the dataset, use `scripts/compute_corrupt.py`.

## Training Script

Uses [Pytorch Lightning](https://github.com/PyTorchLightning/pytorch-lightning) as the boilerplate for reproducibility.

```
python codes/trainer.py --mode train \
    --corrupt_type all \ 
    --batch_size 64 \
    --model_save_dir /checkpoint/koustuvs/dialog_metric/all_dailydialog_10_runs \
    --learn_down True --downsample True --down_dim 300 \
    --optim adam,lr=0.001 --dropout 0.2 --decoder_hidden 200 \ 
    --data_name convai2 \ 
    --data_loc /checkpoint/koustuvs/dialog_metric/convai2_data/ \
    --use_cluster
```

For baselines, add the appropriate flag:

```
--train_baseline [infersent/ruber/bertnli]
```

## Inference Script

```
# CUDA_VISIBLE_DEVICES=0 python codes/inference.py \ 
    --id $MODEL_ID --model_save_dir $MODEL_SAVE_DIR \
    --model_version $VERSION --train_mode nce \ 
    --corrupt_pre $DATA_LOC --test_suffix true_response \ 
    --test_column true_response --results_file "results.jsonl"
```

Outputs the results in a `jsonl` file. To measure human correaltion with [See et al 2019](https://parl.ai/projects/controllable_dialogue/), specify `--human_eval` flag and `--human_eval_file` location.

## Acknowledgements

- Pytorch Lightning - https://github.com/williamFalcon/pytorch-lightning
- HuggingFace Transformers - https://github.com/huggingface/transformers
- FairSeq - https://github.com/pytorch/fairseq
- Liming Vie's RUBER implementation - https://github.com/liming-vie/RUBER
- Pytorch 1.2.0 - https://pytorch.org/
- ParlAI - https://parl.ai/
- See et al 2019 data - https://parl.ai/projects/controllable_dialogue/

## Citation

If our work is useful for your research, consider citing it using the following bibtex:

```
@article{sinha2020maude,
  Author = {Koustuv Sinha and Prasanna Parthasarathi and Jasmine Wang and Ryan Lowe and William L. Hamilton and Joelle Pineau},
  Title = {Learning an Unreferenced Metric for Online Dialogue Evaluation},
  Year = {2020},
  journal = {ACL},
  arxiv = {2005.00583},
  url = {https://arxiv.org/abs/2005.00583}
}
```

## License

This work is CC-BY-NC 4.0 (Attr Non-Commercial Inter.) licensed, as found in the LICENSE file.


CHANGELOG.md
1.0.0 (April 15, 2020)

Initial Public Release of MaUde, an unreferenced metric for online dialog evaluation, to appear in ACL 2020


codes/inference.py
"""
# Copyright (c) Facebook, Inc. and its affiliates.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.
#
"""
# File to run various inferences
import torch
from args import get_args
from logbook.logbook import LogBook
from codes.models import TransitionPredictorMaxPool
from backup.corrupt import CorruptDialog
from data import ParlAIExtractor, id_context_collate_fn
from tqdm import tqdm
import pickle as pkl
from pytorch_lightning import Trainer
from codes.models import TransitionPredictorMaxPool
from codes.baseline_models import RuberUnreferenced, InferSent, BERTNLI
import pandas as pd
import numpy as np
import os.path
import glob
import json
from utils import batchify, batch_dialogs, batch_words


def get_dial_tokens(test_data, X):
    X = [test_data.tokenizer.tokenize(sent) for sent in X]
    X = [test_data.tokenizer.convert_tokens_to_ids(sent) for sent in X]
    return X


def test_trainer(args, model):
    trainer = Trainer(
        gpus=-1,
        show_progress_bar=args.use_cluster == False,
        row_log_interval=10,
        log_save_interval=1,
        distributed_backend="dp",  # if args.debug else 'ddp',
        max_nb_epochs=1,
        # amp_level='O2', use_amp=True,
    )
    trainer.test(model)


def test_single(args, model, df_name, response_row="response", save_eval=False):
    """
    Evaluate the model on a given dataset
    This would be a single batched inference
    """
    df = pd.read_csv(df_name)
    pb = tqdm(total=len(df))
    test_dl = model.get_dataloader(mode="test", datamode="test")
    all_scores = []
    for i, row in df.iterrows():
        context = row["context"]
        response = row[response_row]
        context = context.split("\n")
        X = [model.test_data.tokenizer.tokenize(sent) for sent in context]
        X = [model.test_data.tokenizer.convert_tokens_to_ids(sent) for sent in X]
        Y = model.test_data.tokenizer.convert_tokens_to_ids(
            model.test_data.tokenizer.tokenize(response)
        )
        X_dial_len = None
        if model.bert_input:
            if model.is_transition_fn:
                X = [
                    model.test_data.tokenizer.build_inputs_with_special_tokens(sent)
                    for sent in X
                ]
                X, X_len, X_dial_len = batch_dialogs([X])
            else:
                # flatten
                X = [word for sent in X for word in sent]
                X = model.test_data.tokenizer.build_inputs_with_special_tokens(X)
                X, X_len = batch_words([X])
            Y = model.test_data.tokenizer.build_inputs_with_special_tokens(Y)
        else:
            X = [word for sent in X for word in sent]
            X, X_len = batch_words([X])
        Y, Y_len = batchify([Y], False)
        X_len = torch.from_numpy(X_len)
        Y_len = torch.from_numpy(Y_len)
        if model.is_transition_fn:
            score = model(X, X_len, X_dial_len, Y, Y_len)
        else:
            score = model(X, X_len, Y, Y_len)
        df.at[i, "{}_score_{}".format(args.id, args.model_version)] = score.item()
        all_scores.append(score)
        pb.update(1)
    pb.close()
    mean = np.mean(all_scores)
    std = np.std(all_scores)
    result = "> {} : mean score : {}, std: {}".format(response_row, mean, std)
    cell_format = "{}+/-{}".format(mean, std)
    with open(args.results_file, "a",) as fp:
        res_d = {
            "id": args.id,
            "response_row": response_row,
            "model_version": args.model_version,
            "test_file": df_name,
            "score_mean": str(mean),
            "score_std": str(std),
            "formatted": "{:0.2f}+/-{:0.2f}".format(mean, std),
        }
        fp.write(json.dumps(res_d) + "\n")
    print(result)
    if save_eval:
        print("saving ...")
        df.to_csv(df_name)


def test_context_corruption(args, test_data, model, only_context=False):
    # single non-batched evaluation, may take time
    mode = "ref_score"
    if only_context:
        mode = "cont_score"
    cd = CorruptDialog(args, test_data, False, bert_tokenize=True)
    progress_scores = []
    pb = tqdm(total=len(test_data.dialogs))
    for dial_id, dial in enumerate(test_data.dialogs):
        # only get the last one
        dial_scores = []
        X = dial[0:-1]
        Y = dial[-1]
        X = get_dial_tokens(test_data, X)
        Y = test_data.tokenizer.convert_tokens_to_ids(test_data.tokenizer.tokenize(Y))
        for X_hat in cd.next_corrupt_context_model(dial_id, len(X)):
            X_h = get_dial_tokens(test_data, X_hat)
            batch = id_context_collate_fn([(X, Y, Y, X_h)])
            bX, bX_len, bY, bY_hat, bX_h, bX_h_len = batch
            pred_true = model.forward(bX, bX_len, bY, mode=mode)
            pred_false = model.forward(bX_h, bX_h_len, bY, mode=mode)
            dial_scores.append((pred_true.item(), pred_false.item()))
        progress_scores.append(dial_scores)
        pb.update(1)
    pb.close()
    pkl.dump(progress_scores, open(experiment_path + "/dial_scores_context.pkl", "wb"))


if __name__ == "__main__":
    args = get_args()
    logbook = LogBook(vars(args))
    logbook.write_metadata_logs(vars(args))
    ## load data
    logbook.write_message_logs("init loading data for testing")
    # args.mode = 'test'
    # test_data = ParlAIExtractor(args, logbook)
    # test_data.load()
    # to get the correct paths
    # args = test_data.args
    experiment_path = "{}/{}/lightning_logs/version_{}".format(
        args.model_save_dir, args.id, args.model_version
    )
    model_save_path = "{}/checkpoints/*.ckpt".format(experiment_path)
    all_saved_models = glob.glob(model_save_path)
    model_save_path = all_saved_models[0]
    tag_path = "{}/meta_tags.csv".format(experiment_path)
    # load model
    model_module = None
    if len(args.train_baseline) > 0 and args.train_baseline != "na":
        if args.train_baseline == "ruber":
            model_module = RuberUnreferenced
        elif args.train_baseline == "infersent":
            model_module = InferSent
        elif args.train_baseline == "bertnli":
            model_module = BERTNLI
    else:
        model_module = TransitionPredictorMaxPool
    # load metrics
    model = model_module.load_from_metrics(
        weights_path=model_save_path, tags_csv=tag_path
    )
    # Reset some parameters for easy inference
    model.set_hparam("corrupt_type", args.corrupt_type)
    # model.hparams.corrupt_type = args.corrupt_type
    # set model on evaluation mode
    model.preflight_steps()
    model.eval()
    model.freeze()

    # test_context_corruption(args, test_data, model)
    # test_trainer(args, model)
    if args.human_eval:
        test_file = args.human_eval_file
        test_single(args, model, test_file, save_eval=True)
    else:
        test_file = args.corrupt_pre + args.test_suffix + ".csv"
        if os.path.exists(test_file) and os.path.isfile(test_file):
            print("Testing file : {}".format(test_file))
            test_single(args, model, test_file, response_row=args.test_column)
        else:
            raise FileNotFoundError("File {} not found".format(test_file))


args.py
"""
# Copyright (c) Facebook, Inc. and its affiliates.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.
#
"""
from test_tube import HyperOptArgumentParser
import argparse


def get_args(command=None):
    # parser = argparse.ArgumentParser()
    parser = HyperOptArgumentParser(strategy="random_search", add_help=False)
    # Analysis args
    # parser.add_argument("--bert_model", default='bert-base-uncased', type=str, help="bert model name")
    parser.add_argument(
        "--bert_model",
        default="distilbert-base-uncased",
        type=str,
        help="bert model name",
    )
    parser.add_argument(
        "--emb_file",
        default="~/checkpoint/bert_vectors/",
        type=str,
        help="location of embedding file",
    )
    parser.add_argument(
        "--data_loc",
        default="~/checkpoint/dialog_metric/convai2_data/",
        type=str,
        help="location of data dump",
    )
    parser.add_argument(
        "--data_name", default="convai2", type=str, help="convai2/cornell_movie"
    )
    parser.add_argument(
        "--tok_file", default="na", type=str, help="tokens and word dict file"
    )
    parser.add_argument(
        "--pca_file", default="na", type=str, help="pca saved weights file"
    )
    parser.opt_list(
        "--learn_down",
        default=False,
        action="store_true",
        options=[True, False],
        tunable=False,
    )
    parser.opt_list(
        "--fix_down",
        default=False,
        action="store_true",
        options=[True, False],
        tunable=False,
    )
    parser.add_argument(
        "--trained_bert_suffix",
        default="ep_10_lm",
        type=str,
        help="folder to look for trained bert",
    )
    parser.add_argument("--tc", default=False, action="store_true")
    parser.opt_list(
        "--downsample",
        default=True,
        action="store_true",
        options=[True, False],
        tunable=False,
    )
    parser.opt_list(
        "--down_dim", type=int, default=300, options=[100, 300, 400], tunable=False
    )
    parser.add_argument("--load_fine_tuned", default=True, action="store_true")
    # parser.add_argument("--fine_tune_model", default="~/checkpoint/dialog_metric/cleaned/bert_lm",type=str)
    parser.add_argument(
        "--fine_tune_model",
        default="~/checkpoint/dialog_metric/cleaned/distilbert_lm",
        type=str,
    )
    # Experiment ID
    parser.add_argument("--id", default="ruber_bs", type=str)
    # Model training args
    parser.add_argument("--device", default="cuda", type=str, help="cuda/cpu")
    parser.add_argument(
        "--model",
        default="models.TransitionPredictorMaxPoolLearnedDownsample",
        type=str,
        help="full model name path",
    )
    parser.opt_list(
        "--optim",
        default="adam,lr=0.0001",
        type=str,
        help="optimizer",
        options=["adam,lr=0.001", "adam,lr=0.01", "adam,lr=0.0001"],
        tunable=False,
    )
    parser.add_argument("--epochs", default=10, type=int, help="number of epochs")
    parser.add_argument("--margin", default=0.5, type=float, help="margin")
    parser.add_argument(
        "--train_mode",
        default="ref_score",
        type=str,
        help="ref_score/cont_score/all/nce",
    )
    parser.add_argument(
        "--num_nce", type=int, default=5, help="number of nce samples per scheme"
    )
    parser.add_argument(
        "--model_save_dir",
        default="~/checkpoint/dialog_metric/",
        type=str,
        help="model save dir",
    )
    parser.add_argument(
        "--model_load_path",
        default="~/checkpoint/dialog_metric/",
        type=str,
        help="if there is a need of different load path",
    )
    parser.add_argument("--batch_size", default=64, type=int, help="batch size")
    parser.add_argument(
        "--load_model",
        default=False,
        action="store_true",
        help="load model from previous checkpoint",
    )
    parser.add_argument(
        "--logger_dir",
        default="/private/home/koustuvs/mlp/latentDialogAnalysis/logs/",
        type=str,
        help="batch size",
    )
    parser.add_argument("--log_interval", default=100, type=int, help="log interval")
    parser.add_argument(
        "--watch_model", default=False, action="store_true", help="wandb watch model"
    )
    parser.add_argument(
        "--vector_mode",
        default=True,
        action="store_true",
        help="if false, train with word representations",
    )
    parser.add_argument(
        "--remote_logging",
        default=False,
        action="store_true",
        help="wandb remote loggin on or off",
    )
    parser.add_argument("--wandb_project", default="dialog-metric", type=str)
    parser.add_argument("--bidirectional", default=False, action="store_true")
    parser.add_argument("--dataloader_threads", default=8, type=int)
    parser.add_argument("--exp_data_folder", default="na", help="exp data folder")
    parser.add_argument(
        "--num_workers", default=4, type=int, help="dataloader num workers"
    )
    parser.opt_list(
        "--clip",
        default=0.5,
        type=float,
        help="gradient clipping",
        options=[0.0, 0.5, 1.0],
        tunable=False,
    )
    parser.opt_list(
        "--dropout",
        default=0.2,
        type=float,
        help="gradient clipping",
        options=[0.0, 0.2],
        tunable=False,
    )
    parser.opt_list(
        "--decoder_hidden",
        default=200,
        type=int,
        help="decoder hidden values",
        options=[100, 200, 500, 700],
        tunable=False,
    )
    parser.add_argument(
        "--gpus", type=str, default="-1", help="how many gpus to use in the node"
    )
    parser.add_argument(
        "--debug", default=False, action="store_true", help="if true, set debug modes"
    )
    ## Evaluation args
    parser.add_argument(
        "--corrupt_type",
        default="rand_utt",
        type=str,
        help="all/word_drop/word_order/word_repeat/rand_utt/model_false/rand_back/only_semantics/only_syntax/context_corrupt",
    )
    parser.add_argument(
        "--corrupt_context_type",
        default="rand",
        type=str,
        help="rand/drop/shuffle/model_true/model_false/progress/none",
    )
    parser.add_argument("--drop_per", default=0.50, type=float, help="drop percentage")
    parser.add_argument(
        "--eval_val", default=False, action="store_true", help="only eval val set"
    )
    parser.add_argument(
        "--model_response_pre",
        default="na",
        type=str,
        help="model response file prefix",
    )
    parser.add_argument(
        "--load_model_responses",
        default=True,
        action="store_true",
        help="load model responses",
    )
    parser.add_argument(
        "--corrupt_model_names",
        default="seq2seq",
        type=str,
        help="comma separated models",
    )
    parser.add_argument(
        "--restore_version",
        default=-1,
        type=int,
        help="if > -1, restore training from the given version",
    )

    # Baseline args
    parser.add_argument("--train_baseline", default="na", help="ruber/bilstm", type=str)
    ## RUBER
    parser.add_argument(
        "--word2vec_context_size",
        default=3,
        type=int,
        help="context size for word2vec training",
    )
    parser.add_argument(
        "--word2vec_embedding_dim", default=300, type=int, help="embedding dim"
    )
    parser.add_argument(
        "--word2vec_epochs", default=100, type=int, help="word2vec training epochs"
    )
    parser.add_argument(
        "--word2vec_out",
        default="~/checkpoint/dialog_metric/ruber/w2v.pt",
        type=str,
        help="word2vec output location",
    )
    parser.add_argument("--word2vec_lr", default=0.001, type=float, help="word2vec lr")
    parser.add_argument("--word2vec_batchsize", default=512, type=int)
    parser.add_argument(
        "--ruber_ref_pooling_type", default="max_min", type=str, help="max_min/avg"
    )
    parser.add_argument(
        "--ruber_unref_pooling_type", default="max", type=str, help="max/mean"
    )
    parser.add_argument(
        "--ruber_load_emb", action="store_true", help="load trained word2vec"
    )
    parser.add_argument(
        "--ruber_lstm_dim", default=300, type=int, help="dimensions of ruber encoder"
    )
    parser.add_argument(
        "--ruber_mlp_dim", default=200, type=int, help="dimensions of ruber encoder"
    )
    parser.add_argument(
        "--ruber_dropout", default=0.1, type=float, help="ruber dropout"
    )
    parser.add_argument("--num_words", default=-1, type=int)

    ## Data collection args
    parser.add_argument(
        "--agent", type=str, default="kvmemnn", help="repeat/ir/seq2seq"
    )
    parser.add_argument("--mode", type=str, default="train", help="train/test/valid")
    parser.add_argument(
        "--models",
        type=str,
        default="seq2seq,repeat",
        help="comma separated model values",
    )
    parser.add_argument(
        "--response_file", type=str, default="~/checkpoint/dialog_metric/"
    )
    parser.add_argument(
        "--mf",
        type=str,
        default="/checkpoint/parlai/zoo/convai2/seq2seq_naacl2019_abibaseline/model",
        help="only for special cases",
    )
    parser.add_argument(
        "--only_data",
        action="store_true",
        default=False,
        help="only extract and store dialog data",
    )

    ## SLURM args
    parser.add_argument(
        "--slurm_log_path",
        type=str,
        default="~/checkpoint/dialog_metrics/ckpt/",
        help="slurm log path",
    )
    parser.add_argument(
        "--per_experiment_nb_gpus", type=int, default=1, help="number of gpus"
    )
    parser.add_argument(
        "--per_experiment_nb_cpus", type=int, default=16, help="number of cpus"
    )
    parser.add_argument(
        "--nb_gpu_nodes", type=int, default=1, help="number of gpu nodes"
    )
    parser.add_argument("--job_time", type=str, default="23:59:00", help="time")
    parser.add_argument("--gpu_type", type=str, default="volta", help="gpu type")
    parser.add_argument(
        "--gpu_partition", type=str, default="learnfair", help="gpu type"
    )
    parser.add_argument(
        "--nb_hopt_trials",
        type=int,
        default=1,
        help="how many grid search trials to run",
    )
    parser.add_argument("--train_per_check", type=float, default=1.0)
    parser.add_argument("--val_per_check", type=float, default=1.0)
    parser.add_argument("--test_per_check", type=float, default=1.0)
    parser.add_argument(
        "--use_cluster",
        action="store_true",
        default=False,
        help="activate cluster mode",
    )
    ## Inference args
    parser.add_argument("--model_name", type=str, default="na", help="model name")
    parser.add_argument(
        "--model_version", type=str, default="version_0", help="model version"
    )
    parser.add_argument("--use_ddp", action="store_true", default=False)
    parser.add_argument("--human_eval", action="store_true", default=False)
    parser.add_argument(
        "--human_eval_file",
        type=str,
        default="~/checkpoint/dialog_metric/controllable_dialogs.csv",
    )
    parser.add_argument("--results_file", type=str, default="test_results.jsonl")
    ## Corruption args
    parser.add_argument(
        "--corrupt_pre",
        type=str,
        default="~/checkpoint/dialog_metric/convai2_data/convai2_test_",
    )
    parser.add_argument("--corrupt_ne", type=int, default=1)
    parser.add_argument("--test_suffix", type=str, default="true_response")
    parser.add_argument("--test_column", type=str, default="true_response")

    if command:
        return parser.parse_args(command.split(" "))
    else:
        return parser.parse_args()
